### PR TITLE
*: add "--unsafe-allow-cluster-version-downgrade" for not failing cluster version downgrade

### PR DESCRIPTION
**Pull request by chaochn47**
Compared to the [public etcd downgrade design](https://docs.google.com/document/d/1mSihXRJz8ROhXf4r5WrBGc8aka-b8HKjo2VDllOc6ac/edit#heading=h.e4jdx621yd8s), I am trying to simplify the manual "whitelisting target downgrade version" process and it doesn't require any server/client side API changes. Once the new `--unsafe-allow-cluster-version-downgrade` flag is enabled, only one minor version downgrade is allowed (e.g. v3.6 to v3.5). We can cherry-pick the same change to v3.4 and v3.3 if this feature is useful. 

However, if a user is already using `--experimental-enable-lease-checkpoint` flag to bootstrap a v3.4 cluster, the version downgrade is not possible due to the  [MustUnmarshal](https://github.com/etcd-io/etcd/blob/release-3.3/etcdserver/server.go#L1436-L1442) will panic on `LeaseCheckpoint` internal raft message which doesn't exist in v3.3. Not to mention there is no corresponding apply method. As a result, the new added etcd v3.3 flag `--unsafe-allow-cluster-version-downgrade` is not expected to set to `true` in this case. 

I've excessively tested v3.5 to v3.4 and v3.4 to v3.3 downgrade in two environments and the results look good. I will publish the backport PRs once the community agree on the approach. 

1. Tested locally first bootstrap with 3 etcd v3.4 servers and then downgrade to v3.3 one by one. It aims to test when the v2 store cluster version request replicate from v3.4 to new v3.3 server, it can be replaced by local 3.3 cluster version. 
2. Tested in a 1.17 kubernetes cluster which by default have 3 etcd v3.4 servers. It amis to test new v3.3 server can recover from the snapshot and replace the cluster version. At the same time when downgrading starts, I ran the sonobuoy e2e conformance test to verify it won't cause k8s functionality regression. 

I understand this is a risky operation without controlling the set of APIs used by the etcd client. But, I would like to put it here for reference, as others may find it useful.

Any feedback is welcome, Thanks!

### CODE REVIEW

**Comment by gyuho** [comment]
@jpbetz Do we have any idea how widely lease checkpointer is used? The main motivation here is, while downgrade API help you work around allowing version checks, it still does not prevent crashes from new message field like lease checkpointer. Then why don't we just allow built-in downgrade, for whomever is willing to take the risk and the risk can be calculated if one controls the client behavior.

**Review by lilic — COMMENTED** [review]
This makes sense to me, but I was wondering could also work together with the downgrade subcommand, should we then maybe put it under that as a flag? 
`etcdctl downgrade --unsafe`

Note the downgrade command work still needs to finished up. I just wonder if that is a somewhat better UX?

**Comment by chaochn47** [comment]
A big reason for adding a flag is the downgrade api is not completed in client side and we need to integrate this to the operator downgrade workflow.

So this PR provides a simpler way for minor version downgrade automation.

File: server/embed/config.go [hunk 1]
```diff
@@ -489,7 +499,8 @@ func NewConfig() *Config {
 		ExperimentalMemoryMlock:                  false,
 		ExperimentalTxnModeWriteWithSharedBuffer: true,
 
-		V2Deprecation: config.V2_DEPR_DEFAULT,
+		UnsafeAllowClusterVersionDowngrade: DefaultUnsafeAllowClusterVersionDowngrade,
+		V2Deprecation:                      config.V2_DEPR_DEFAULT,
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply server/embed/config.go:503]: Good point, will fix it.

**Comment by lilic** [comment]
I added downgrade API to the client, but the part of the command still need to be worked on, I started the discussion on original issue https://github.com/etcd-io/etcd/issues/11716.

**Pull request closed without merging**